types/src/error.rs

```diff
@@ -177,32 +177,44 @@ impl Error {
 	}
 }
 
-/// Error type with a special `subscription_closed` field to detect that
+/// A type with a special `subscription_closed` field to detect that
 /// a subscription has been closed to distinguish valid items produced
 /// by the server on the subscription stream from an error.
 #[derive(Deserialize, Serialize, Debug, PartialEq)]
-pub struct SubscriptionClosedError {
-	subscription_closed: String,
-	id: u64,
+pub struct SubscriptionClosed {
+	subscription_closed: SubscriptionClosedReason,
 }
 
-impl SubscriptionClosedError {
-	/// Create a new subscription closed error.
-	pub fn new(reason: impl Into<String>, id: u64) -> Self {
-		Self { subscription_closed: reason.into(), id }
+impl From<SubscriptionClosedReason> for SubscriptionClosed {
+	fn from(kind: SubscriptionClosedReason) -> Self {
+		Self::new(kind)
 	}
+}
 
-	/// Get the reason why the subscription was closed.
-	pub fn close_reason(&self) -> &str {
-		&self.subscription_closed
+impl SubscriptionClosed {
+	/// Create a new subscription message.
+	pub fn new(kind: SubscriptionClosedReason) -> Self {
+		Self { subscription_closed: kind }
 	}
 
-	/// Get the subscription ID.
-	pub fn subscription_id(&self) -> u64 {
-		self.id
+	/// Get the close reason.
+	pub fn close_reason(&self) -> &SubscriptionClosedReason {
+		&self.subscription_closed
 	}
 }
 
+/// A type to represent when a subscription gets closed
+/// by either the server or client side.
+#[derive(Deserialize, Serialize, Debug, PartialEq)]
+pub enum SubscriptionClosedReason {
+	/// The subscription was closed by calling the unsubscribe method.
+	Unsubscribed,
+	/// The client closed the connection.
+	ConnectionReset,
+	/// The server closed the subscription.
+	Server(String),
```

Comment:
Would be good to document what information the `String` carries here.

Reply:
Sure, we got the [SubscriptionSink::close](https://github.com/paritytech/jsonrpsee/blob/master/utils/src/server/rpc_module.rs#L807) where one can add the reason why a subscription was closed on the server-side.